### ISSUE TITLE
provide AppImage release for arm / aarch64

### ISSUE THREAD

**Issue by goshhhy** [neovim/neovim#15143]
- `nvim --version`: N/A
- Operating system/version: Debian testing, arm64

### Actual behaviour

there is no AppImage for aarch64 platforms. this means users like me who do not use x86_64 and run a distribution without the latest version in its repositories, will have to build it ourselves if we want the latest.

### Expected behaviour

an AppImage for aarch64 is provided.

**Comment by clason** [comment]
I believe that is blocked by not having ARM64 runners for GH Actions: https://github.com/actions/virtual-environments/issues/2552

**Comment by rktjmp** [comment]
Looks like it's stuck as a wontfix for now too...

One solution is a self hosted GH runner, AWS has ARM instances, as well as a few others here and there.

I imagine this would need a benevolent user to donate an instance to happen though, and it's open to abuse by malicious PRs.

Also it's probably a nightmare to keep upright.

https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners.

**Comment by bfredl** [comment]
Well we could use such a thing only for nightly builds, and not for PR runs (whose queue probably is going to be impossible anyway, unless we spinn up a lot of instances)

**Comment by ck3mp3r** [comment]
Why don't you use docker buildx? 
I'm able to compile and build docker images with neovim for both amd64 and arm64 using GitHub actions
I think the trick here is QUEMU...

**Comment by rktjmp** [comment]
Could you post an example action & docker file, if you have one already that is, save some ground work.

**Comment by ck3mp3r** [comment]
Yeah, build time for amd64 and arm64 containers together take just over an hour... but still better than having nothing for arm64, especially as there are going to be more of them around...

**Comment by rjahrj** [comment]
Is this why I cant use the App Image? or the tar package. I get a "Exec format err" for both. I believe i am having an issue with the Ubuntu package(at least i think i am) the unstable build is version 5.01Lua and not LuaJIT. Unless this is not a problem. I am having issues with LuaRocks not being present, its affecting packer.nvim. Or im stupid and have no idea what I am talking about. I just want to program Svelte lol but I cant seem to get my LSP right lol.

**Comment by rktjmp** [comment]
Just noting down WIP here: https://github.com/rktjmp/neovim/tree/ci-arm

Don't use the released files in that repo yet, they are a mix of arm64 and amd64 because arm builds too slow for testing.

**Comment by rktjmp** [comment]
Draft PR https://github.com/neovim/neovim/pull/15542 which needs some testing by actual users.

**Comment by tigerinus** [comment]
With the understanding that there might not be a ARM device for building the ARM package, is cross-compiling something open for discussion here?

There are many aged ARM devices re-enabled for regular work scenarios (e.g. thin-terminal, IoT controller, file/printer sharing, etc.) by Linux.

**Comment by rktjmp** [comment]
The PR above would support any platform QEMU does, just have to add the arch to the matrix. That may be simpler than setting up cross compilation toolchains in "normal" builds.

I imagine the core team would have to weigh up potential support costs (issue triage, platform quirks, etc) if they were to provide separate binaries for many platforms, but an industrious user could probably:

- fork neovim, update the builder with whatever platforms they wanted and create a new action to rebase those changes onto the latest HEAD every day and act as a 3rd party _unsupported, unofficial_ release provider, or
- use the dockerfiles to build whatever platform they wanted locally.

**Comment by remiconnesson** [comment]
> Looks like it's stuck as a wontfix for now too...
> [...]
> I imagine this would need a benevolent user to donate an instance to happen though, and it's open to abuse by malicious PRs.

Maybe that will help: https://www.oracle.com/cloud/free/ 

Oracle Cloud provide a Free Tier which include an instance always free with an ARM CPU and 24GB of RAM

**Comment by krmbzds** [comment]
> I believe that is blocked by not having ARM64 runners for GH Actions: [actions/runner-images#2552](https://github.com/actions/runner-images/issues/2552)

GitHub ARM64 runner issue is now tracked here: https://github.com/actions/runner-images/issues/5631

**Comment by hardboydu** [comment]
https://github.com/hardboydu/neovim-build

build on FT D2000 CentOS 7.x

**Comment by afifurrohman-id** [comment]
> > I believe that is blocked by not having ARM64 runners for GH Actions: [actions/runner-images#2552](https://github.com/actions/runner-images/issues/2552)
> 
> GitHub ARM64 runner issue is now tracked here: [actions/runner-images#5631](https://github.com/actions/runner-images/issues/5631)

@pentamassiv Now MacOS M1 is GA: https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source

**Comment by clason** [comment]
> Sadly, M1 doesn't support nested virtualization due to architectural issue:
https://github.com/actions/runner-images/issues/9460

Yes, and that is why we are still blocked. No native ARM linux runners == no official native ARM linux releases.

**Comment by afifurrohman-id** [comment]
> > Sadly, M1 doesn't support nested virtualization due to architectural issue:
> > [actions/runner-images#9460](https://github.com/actions/runner-images/issues/9460)
> 
> Yes, and that is why we are still blocked. No native ARM linux runners == no official native ARM linux releases.

I think Circle CI provide it ;)

**Comment by xnzac** [comment]
Just to chime in, to build on your own machine(tested on Ubuntu 22.04 via UTM on Mac M1):

1) sudo apt-get install ninja-build \
     gettext libtool libtool-bin \
     autoconf automake cmake g++ \
     pkg-config unzip

2) git clone https://github.com/neovim/neovim.git

3) git checkout stable

4) sudo make CMAKE_BUILD_TYPE=Release \
     CMAKE_INSTALL_PREFIX=<NEOVIM_LOCATION> install

I used /opt/nvim for <NEOVIM_LOCATION>

5) export PATH=”<NEOVIM_LOCATION>/bin:$PATH”

Credit: https://carlosahs.medium.com/how-to-install-neovim-from-source-on-ubuntu-20-04-lts-524b3a91b4c4

**Comment by Spuxy** [comment]
@xnzac i get that point, but again, u have tested on __Mac M1__. Try on rpi or some linux based os with cpu arch of arm.
+ If you are going to build that by their doc `BUILD.md`, it would be simple (i was compiling but i lost interested to finish that :) ) even for arm64 on unix-like systems.

Link on their official [doc](https://github.com/neovim/neovim/blob/master/BUILD.md#building)

**Comment by mpriscella** [comment]
ARM based instances are now in [public beta](https://github.blog/changelog/2024-06-03-actions-arm-based-linux-and-windows-runners-are-now-in-public-beta/). Hopefully this unblocks this issue now.

**Comment by pidgeon777** [comment]
> ARM based instances are now in [public beta](https://github.blog/changelog/2024-06-03-actions-arm-based-linux-and-windows-runners-are-now-in-public-beta/). Hopefully this unblocks this issue now.

Let's hope! 🤞

**Comment by clason** [comment]
> ARM based instances are now in [public beta](https://github.blog/changelog/2024-06-03-actions-arm-based-linux-and-windows-runners-are-now-in-public-beta/). Hopefully this unblocks this issue now.

As always, these beta offerings only apply to the paid-for, non-open source, tiers. So this issue remains blocked for the foreseeable future.

(But GH eventually released ARM-based macOS runners to the general public, so there's hope -- when Microsoft starts dropping Intel so ARM will become the standard architecture for Windows.)

**Comment by BitRanger** [comment]
> So this issue remains blocked for the foreseeable future.


The Github Actions post states that:
> Arm runners are available to customers on our Team and Enterprise Cloud plans. We expect to begin offering Arm runners for open source and personal accounts by the end of the year.

So hopefully the a runner for the arm builds can be created around that time, assuming it becomes availible to neovim.

**Comment by benz0li** [comment]
In the meantime: https://gitlab.b-data.ch/neovim/neovim/-/releases

❗ *Unofficial* and *untested* builds. No intention to build AppImages.

**Comment by benz0li** [comment]
ℹ️ Multi-arch (`linux/amd64`, `linux/arm64/v8`) docker images: https://gitlab.b-data.ch/neovim/nvsi/container_registry.

Usage:

```Dockerfile
FROM glcr.b-data.ch/neovim/nvsi:nightly AS nvsi

FROM debian:12

COPY --from=nvsi /usr/local /usr/local
```

👉 `neovim/nvsi` images simply store a Neovim installation at `/usr/local` – you cannot use these to run `nvim` itself.

**Comment by clason** [comment]
No, these are not publicly available (for open source projects; Enterprise only).

**Comment by Tanq16** [comment]
Offering an ARM64 .deb is also not a bad option until native ARM runners are available for public. I build it this way for my personal ARM64 containers -

```
FROM ubuntu:jammy AS executable_builder
RUN mkdir /executables && apt update -y && \
    apt install -y wget ninja-build gettext cmake unzip curl git file && \
    wget https://github.com/neovim/neovim/archive/refs/tags/stable.tar.gz && \
    tar -xvf stable.tar.gz && cd neovim-stable && \
    make CMAKE_BUILD_TYPE=RelWithDebInfo && \
    cd build && cpack -G DEB && \
    mv nvim-linux64.deb /
```

**Comment by clason** [comment]
Linux arm64 runners are now available for free public repositories, so this can be looked at. Note that the runners use a limited `ubuntu-24.04` image so not all tooling required for our workflows are available, so it might not be as easy as slapping a `run-on:` on them.

We will investigate this; no need to add comments here. @dundargoc

**Comment by clason** [comment]
Be my guest.
